ADAL/src/cache/ADAuthenticationContext+TokenCaching.m

```diff
@@ -188,14 +191,17 @@ - (void)updateCacheToResult:(ADAuthenticationResult*)result
             }
             
             BOOL removed = NO;
-            //The refresh token didn't work. We need to clear this refresh item from the cache.
+            //The refresh token didn't work. We need to tombstone this refresh item in the cache.
             ADTokenCacheKey* exactKey = [cacheItem extractKey:nil];
             if (exactKey)
             {
                 ADTokenCacheItem* existing = [tokenCacheStoreInstance getItemWithKey:exactKey userId:cacheItem.userInformation.userId error:nil];
                 if ([refreshToken isEqualToString:existing.refreshToken])//If still there, attempt to remove
                 {
-                    AD_LOG_VERBOSE_F(@"Token cache store", [self correlationId], @"Removing cache for resource: %@", cacheItem.resource);
+                    AD_LOG_VERBOSE_F(@"Token cache store", requestCorrelationId, @"Tombstoning cache for resource: %@", cacheItem.resource);
+                    //set request correlation ID and error details in tombstone property
+                    [existing setTombstone:[NSMutableDictionary dictionaryWithDictionary:@{ @"correlationId" : [requestCorrelationId UUIDString],
+                                                                                            @"errorDetails" : [result.error errorDetails]}]];
                     [tokenCacheStoreInstance removeItem:existing error:nil];
```

Comment:
By problem you mean tests are failing? We should update the tests to expect the tombstones to be there might need to update tests to not count tombstones. We also could update allItems to not return tombstones. I'm fine with either.